c++/src/kj/compat/http.c++

```diff
@@ -6231,80 +6341,91 @@ private:
-        fulfiller->fulfill(HttpClient::ConnectResponse {
-          HttpClient::ConnectResponse::Status {
-            statusCode,
-            statusTextCopy,
-            headersCopy.get(),
-            kj::mv(body),
-          },
-          heap<NullIoStream>()
-        });
-      }).eagerlyEvaluate([](kj::Exception&& e) { KJ_LOG(ERROR, e); });
-
-      return kj::mv(pipe.out);
+      respond(statusCode, statusText, headers);
+      kj::get<2>(streamAndFulfiller)->reject(
+          KJ_EXCEPTION(DISCONNECTED, "the tunnel request was rejected"));
+      return getRejectionStream();
     }
 
   private:
-    kj::Own<kj::PromiseFulfiller<HttpClient::ConnectResponse>> fulfiller;
-    kj::Promise<void> task = nullptr;
+    using StreamsAndFulfiller =
+        kj::Tuple<kj::Own<kj::AsyncIoStream>,
+                  kj::Own<kj::AsyncIoStream>,
+                  kj::Own<kj::PromiseFulfiller<void>>>;
+    // The first stream is the original owned stream.
+    // The second stream is the wrapped/guarded stream that wrapps a reference to
+    // the first. This will be handed off when getTunnelStream() is called.
+    // The fulfiller is used to resolve the guard for the second stream. This will
+    // be fulfilled or rejected when accept/reject is called.
+
+    kj::Own<kj::PromiseFulfiller<HttpClient::ConnectRequest::Status>> fulfiller;
+    StreamsAndFulfiller streamAndFulfiller;
+    kj::Maybe<kj::Own<kj::HttpHeaders>> headersCopy;
+    kj::Maybe<kj::String> statusTextCopy;
+    bool tunnelStreamDetached = false;
+
+    StreamsAndFulfiller initStreamsAndFulfiller(kj::Own<kj::AsyncIoStream> stream) {
+      auto paf = kj::newPromiseAndFulfiller<void>();
+      auto wrapped = kj::Own<kj::AsyncIoStream>(stream.get(), kj::NullDisposer::instance);
+      auto guarded = kj::heap<AsyncIoStreamWithGuards>(
+          kj::mv(wrapped),
+          kj::Maybe<HttpInputStreamImpl::ReleasedBuffer>(nullptr),
+          kj::mv(paf.promise)).attach(kj::addRef(*this));
+      return kj::tuple(
+        kj::mv(stream),
+        kj::mv(guarded),
+        kj::mv(paf.fulfiller));
+    }
+
+    kj::Own<kj::AsyncIoStream> getTunnelStream() {
+      KJ_ASSERT(!tunnelStreamDetached, "the tunnel stream was already detached");
+      tunnelStreamDetached = true;
+      return kj::mv(kj::get<1>(streamAndFulfiller));
+    }
+
+    kj::Own<kj::AsyncOutputStream> getRejectionStream() {
+      auto out = kj::Own<kj::AsyncOutputStream>(
+          kj::get<0>(streamAndFulfiller).get(), kj::NullDisposer::instance);
+      return out.attach(kj::addRef(*this));
+    }
+
+    void respond(uint statusCode, kj::StringPtr statusText, const HttpHeaders& headers) {
+      KJ_REQUIRE(statusTextCopy == nullptr, "the connect request already has a response");
+      statusTextCopy = kj::str(statusText);
+      headersCopy = kj::heap(headers.clone());
+      fulfiller->fulfill(HttpClient::ConnectRequest::Status {
+        statusCode,
+        KJ_ASSERT_NONNULL(statusTextCopy),
```

Comment:
Agreed. Do you have a preference for either (a) or (b)? I think I'd be good with either.

Reply:
Went ahead and implemented (a).

Reply:
So, for HTTP requests, the idea behind making the headers non-owned is that the headers object would actually be able to point directly into HttpOutputStream's buffer, so the header content never needed to be copied.

I guess that optimization might be difficult to achieve with `CONNECT` when pipelining is supported, though, since the buffer is discarded on the first read, which the client might want to issue before the status promise even resolves. So maybe it's for the best to just not attempt that...

Reply:
I think there might be a path to getting there but it's likely not worth spinning wheels on it right now. Just having `Status` own those right now is straightforward. We can revisit this later. I'll add a TODO comment